### DISCUSSION TITLE
My LCD Screen is constantly white

### DISCUSSION THREAD

**Discussion by BrandonStudio** [Bodmer/TFT_eSPI#2071] [Q&A - General]
I'm using ESP32-WROOM-32E and an 8-bit MCU LCD. The LCD is driven by ST7789V and has white led back light.
The hardware connection is like this:
<img src="https://user-images.githubusercontent.com/55647556/195036090-e67163fb-1720-43be-86b3-f710bde91396.png" width="320">

I've correctly set macros to fit this lcd via `platformio.ini` and passed `TFT_User_Setup` Test with following output:
```
[code]
TFT_eSPI ver = 2.4.77
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = Parallel
Display driver = 7789
Display width  = 240
Display height = 320

TFT_CS   = GPIO 27
TFT_DC   = GPIO 33
TFT_WR   = GPIO 32

TFT_D0   = GPIO 12
TFT_D1   = GPIO 16
TFT_D2   = GPIO 17
TFT_D3   = GPIO 18
TFT_D4   = GPIO 19
TFT_D5   = GPIO 21
TFT_D6   = GPIO 5
TFT_D7   = GPIO 15

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

[/code]
```
One minor problem is that before these it said:
```
[    16][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
E (13) gpio: gpio_set_level(226): GPIO output gpio_num error
```
I think it may be the problem that `TFT_RD` is defiened as `-1`, but it is both useless either adding `-UTFT_RD` in `build_flags` or adding `#undef TFT_RD` in source file.

`TFT_ReadWrite_Test` failed with following output:
```
[    16][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
E (13) gpio: gpio_set_level(226): GPIO output gpio_num error
 Pixel value written = 1
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 2
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 4
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 8
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 10
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 20
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 40
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 80
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 100
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 200
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 400
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 800
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 1000
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 2000
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 4000
 Pixel value read    = 76B5
 ERROR                 ^^^^
 Pixel value written = 8000
 Pixel value read    = 76B5
 ERROR                 ^^^^
```

I saw https://github.com/Bodmer/TFT_eSPI/issues/1932#issuecomment-1199003093 where @gjssss said that it may have something to do with backlight pin ad CS pin, but I'm not fully understand.

Could someone please help me? I'd like to know what the problem is, is it because the difference between ST7789 and ST7789V or because of the white backlight or because of the wrong wiring?

**Comment by Bodmer** [comment]
The CS signal is active low, so you must tie LCD_CS low, i.e. to 0V (GND).

You have not connected the RD pin to the processor so reading pixel data will not work!

**Reply by BrandonStudio** [reply]
But I'd specify the `TFT_CS` to GPIO27, do you mean that I have to `pinMode(27, PULLUP)` and `digitalWrite(27, LOW)` manually before calling `TFT.init()`?

**Reply by BrandonStudio** [reply]
But this still doesn't work.
Or should I remove the pull-up resistor?

**Reply by Bodmer** [reply]
Ok, my mistake I had not noticed the pin was allocated, I just looked at the schematic. The pullup can be left in place.

The DC and WR pins must be allocated GPIO numbers in the range 0-31, see example here and note comments:
https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setups/Setup13_ILI9481_Parallel.h

**Reply by BrandonStudio** [reply]
@Bodmer I cannot change the pin definitation for a short time, could you please give me a guide about how to modify this library to fit my requirement (using GPIO 32 and 33)? This is rather urgent for me

**Reply by Bodmer** [reply]
I think this will work if you modify TFT_eSPI_ESP32.h:

```
    #if (TFT_DC >= 0) &&  (TFT_DC < 32)
        #define DC_C GPIO.out_w1tc = (1 << TFT_DC)
        #define DC_D GPIO.out_w1ts = (1 << TFT_DC)
   #elif (TFT_DC >= 31)
        #define DC_C GPIO.out1_w1tc.val = (1 << (TFT_DC- 32))
        #define DC_D GPIO.out1_w1ts.val = (1 << (TFT_DC- 32))
    #else
        #define DC_C
        #define DC_D
    #endif
```

**Reply by Bodmer** [reply]
TFT_WR should be OK as GPIO 32 you will just be sacrificing a bit of performance.

**Reply by BrandonStudio** [reply]
@Bodmer this still doesn't work T_T

**Reply by Bodmer** [reply]
There must be an additional problem somewhere. As far as I can tell from the information you have given, it should work correctly now.

What are you connecting the RST line to?
The display may need RST to be toggled then held high.
Have you tried allocating and using a TFT_RST pin?
Have you configured the display correctly to work in 8 bit parallel mode?
Are you keeping connections short to prevent ground bounce?
Does the display flicker at all when the code is run?
Is your power supply adequate?
Are you sure all the connections are correct?
Have you looked at the lines with an oscilloscope to see if they are all toggling?
Does the graphics test run and report on the serial monitor window reasonable numbers?
Do you have a full schematic of the display interface to examine?

**Comment by TangerineDreamer** [comment]
I think you can drop away all the resistances,connect CS to GND,even if it is declared to GPIO27 in the setup,don't need the pinMode
 stuff , connect LEDA directly to 3,3V,and RD to an other PIN of the ESP32.

**Comment by BrandonStudio** [comment]
I noticed that in `TFT_eSPI_ESP32.h`:
```c
#if defined (TFT_PARALLEL_8_BIT)
    // TFT_DC, by design, must be in range 0-31 for single register parallel write
    #if (TFT_DC >= 0) &&  (TFT_DC < 32)
        #define DC_C GPIO.out_w1tc = (1 << TFT_DC)
        #define DC_D GPIO.out_w1ts = (1 << TFT_DC)
    #else
        #define DC_C
        #define DC_D
    #endif
#else
...
#endif
```
But I set DC to GPIO33 and WR to GPIO32, is this the problem?

**Comment by TangerineDreamer** [comment]
Yes it could be, i remember RST can be wired to GPIO32 or 33 ,but TFT_D0 ... TFT_D7 can be wired up to GPIO31, the same
for DC/RS  and WR ,also you have many pins unused below GPIO31.
Look at my setup for a SSD1963 and ESP WROOM,maybe even if it's not the same driver as you ,the ESP32 pins are the same:

//-------------------SSD1963 & ESP32------------------------

#define TFT_PARALLEL_8_BIT



#define SSD1963_800ALT_DRIVER // Untested

#define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red




#define TFT_MISO 19
#define TFT_MOSI 23
#define TFT_SCLK 18


//#define TOUCH_CS 5     // Chip select pin (T_CS) of touch screen





#define TFT_CS   22  //or  -->(GND) Chip select control pin (library pulls permanently low
#define TFT_DC   15  // DC=RS Data Command control pin - must use a pin in the range 0-31
#define TFT_RST  32  // Reset pin, toggles on startup

#define TFT_WR    4  // Write strobe control pin - must use a pin in the range 0-31
#define TFT_RD    2  // Read strobe control pin

#define TFT_D0   12  // Must use pins in the range 0-31 for the data bus
#define TFT_D1   13  // so a single register write sets/clears all bits.
#define TFT_D2   26  // Pins can be randomly assigned, this does not affect
#define TFT_D3   25  // TFT screen update performance.
#define TFT_D4   21
#define TFT_D5   5
#define TFT_D6   27
#define TFT_D7   14



#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts


#define SMOOTH_FONT



#define SPI_FREQUENCY  20000000

#define SPI_READ_FREQUENCY  20000000


#define SPI_TOUCH_FREQUENCY  2500000

**Reply by Bodmer** [reply]
DC must be allocated a pin in range GPIO 0-31
WR should be allocated a pin in the range GPIO 0-31 for best performance.

**Comment by BrandonStudio** [comment]
@Bodmer 

> What are you connecting the RST line to?

I'm connecting to GPIO 0.

> The display may need RST to be toggled then held high.

`digitalRead(0)` tells it is 1.

> Have you tried allocating and using a TFT_RST pin?

Yes, nothing changes.

> Have you configured the display correctly to work in 8 bit parallel mode?

My display only provides parallel MCU interface.
The display is called ZSX024-B2441, a component in SiPEED Maix Dock. The display is not broken because it works on that board.

> Are you keeping connections short to prevent ground bounce?

I think so, but I couldn;t be very sure.

> Does the display flicker at all when the code is run?

No. The display is always white.

> Is your power supply adequate?

Yes.

> Are you sure all the connections are correct?

It should be. I'd tested short and open on my board.

> Have you looked at the lines with an oscilloscope to see if they are all toggling?

No. I just placed an FPC connector on my board, didn't add a header for these pins.

> Does the graphics test run and report on the serial monitor window reasonable numbers?

What is a graphics test?

> Do you have a full schematic of the display interface to examine?

It may be somehow complicated.
![image](https://user-images.githubusercontent.com/55647556/195475189-b52519a1-fd27-4a92-987c-0cb95f91ba12.png)
The PCB may not be available now as I've modified it.
Right now I just place ESP32, 1117, CH340, MicroUSB, 2 NPN, 2 button (RST and EN) and related caps and resistors, so other chips are irrelative.

**Reply by BrandonStudio** [reply]
BTW TxD and RxD connected wronly in the picture but I've changed it.

**Reply by Bodmer** [reply]
This output posted earlier indicates that the TFT_RST pin has not been allocated to GPIO 0 in the setup file:
```
[code]
TFT_eSPI ver = 2.4.77
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = Parallel
Display driver = 7789
Display width  = 240
Display height = 320

TFT_CS   = GPIO 27
TFT_DC   = GPIO 33
TFT_WR   = GPIO 32

TFT_D0   = GPIO 12
TFT_D1   = GPIO 16
TFT_D2   = GPIO 17
TFT_D3   = GPIO 18
TFT_D4   = GPIO 19
TFT_D5   = GPIO 21
TFT_D6   = GPIO 5
TFT_D7   = GPIO 15

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

[/code]
```
 Post a copy of the setup file.

**Reply by BrandonStudio** [reply]
Yes, at that time I didn't define the macro.

**Reply by BrandonStudio** [reply]
Current output:
```
[code]
TFT_eSPI ver = 2.4.77
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = Parallel
Display driver = 7789
Display width  = 240
Display height = 320

TFT_CS   = GPIO 27
TFT_DC   = GPIO 33
TFT_RST  = GPIO 0
TFT_WR   = GPIO 32

TFT_D0   = GPIO 12
TFT_D1   = GPIO 16
TFT_D2   = GPIO 17
TFT_D3   = GPIO 18
TFT_D4   = GPIO 19
TFT_D5   = GPIO 21
TFT_D6   = GPIO 5
TFT_D7   = GPIO 15

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

[/code]
```

**Reply by BrandonStudio** [reply]
I'm using `platformio.ini` to setup
```ini
build_flags =
  -Os
  -DCORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_VERBOSE
  -DUSER_SETUP_LOADED=1
  
  ; Define the TFT driver, pins etc here:
  -DTFT_PARALLEL_8_BIT=1
  -DST7789_DRIVER=1
  -DTFT_WIDTH=240
  -DTFT_HEIGHT=320
  -DTFT_CS=27
  ;-UTFT_CS
  -DTFT_DC=33
  -DTFT_WR=32
  -DTFT_D0=12
  -DTFT_D1=16
  -DTFT_D2=17
  -DTFT_D3=18
  -DTFT_D4=19
  -DTFT_D5=21
  -DTFT_D6=5
  -DTFT_D7=15
  -DTFT_RST=0
  -UTFT_RD
  ;-DTFT_BL=21
  ;-DTOUCH_CS=22
  -DLOAD_GLCD=1
  -DLOAD_FONT2=1
  -DLOAD_FONT4=1
  -DLOAD_FONT6=1
  -DLOAD_FONT7=1
  -DLOAD_FONT8=1
  -DLOAD_GFXFF=1
  -DSMOOTH_FONT=1
  -DSPI_FREQUENCY=27000000
```

**Reply by Bodmer** [reply]
I see the RST line is connected to a capacitor and switch and has no pullup. This means driving from a GPIO pin is not a good idea as the switch will short the CPU output when pressed. Ideally the GPIO 0 should have a series resistor to limit the current into the switch and capacitor if it is going to be driven.

Note that at boot GPIO 0 outputs a PWM signal. I don't think this matters for the display but the processor GPIO will be stressed by that capacitor. See here:
https://randomnerdtutorials.com/esp32-pinout-reference-gpios/

There is a sketch called TFT_graphicstest_one_lib that I use to test displays.

I have checked my notes and see I have never used an ST7789V in parallel mode but I see no reason why it won't work as parallel mode has been tested with many other displays.

**Reply by BrandonStudio** [reply]
EN and RST signal are driven by DTR and RTS with 2 10K resistors and 2 NPN, which is same as official devkit. Do you think it is the problem of the RST signal? Should I just tied it up or hang it?

> https://randomnerdtutorials.com/esp32-pinout-reference-gpios/

Pins there are for ESP32-WROOM-32, not for ESP32-WROOM-32E. Since 32D, number of pin and some pin definations have changed.

`TFT_graphicstest_one_lib` check result:
```
TFT_eSPI library test!
[    16][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
E (13) gpio: gpio_set_level(226): GPIO output gpio_num error
Benchmark                Time (microseconds)
Screen fill              123931
Text                     8789
Lines                    50264
Horiz/Vert Lines         12735
Rectangles (outline)     8120
Rectangles (filled)      317683
Circles (filled)         37837
Circles (outline)        16712
Triangles (outline)      12398
Triangles (filled)       106754
Rounded rects (outline)  12176
Rounded rects (filled)   322374
Done!
```

**Reply by Bodmer** [reply]
I think RST is not the problem.
I am not familiar with the 32E so do not know what pin constraints may exist.
Yes, the graphicstest seriual output timings are good.

I have run out of ideas now...

**Reply by BrandonStudio** [reply]
T_T

**Comment by TangerineDreamer** [comment]
Did you wired LEDA direcly to 3,3V and verify if there is 3,3V between LEDA(anode) and LEDK(cathode)?
Maybe you can lower the clock of the esp32

**Reply by BrandonStudio** [reply]
> Did you wired LEDA direcly to 3,3V and verify if there is 3,3V between LEDA(anode) and LEDK(cathode)?

I haven't tested it yet, but I believe there is, because the GND has been tested to be correct, and the LCD wouldn't be on (white) if LEDA was not connected to 3.3V (when removed R5).

> Maybe you can lower the clock of the esp32

Could you give me some suggestions on how to do that?

**Comment by TangerineDreamer** [comment]
In the TFT setup,down page,you have frequencies you can define:

#define SPI_FREQUENCY 20000000
#define SPI_READ_FREQUENCY 20000000

I've seen on a forum someone lower to 1000000 to correct colors,maybe you can test it, i see your platformio
shows : DSPI_FREQUENCY=27000000.,maybe it's too much ,maybe down to 20000000 or 15MHZ,10MHZ

On your wiring , i see IOVCC wired to VCC,what is this? A potentiometer stuff aim to control the lighting?

You could also show us pictures or http adress of the seller page where you bought it the TFT
Is your esp32 wroom 32e a CME component or a dev kit?

Is your alimentation support the tft consommation, wich often goes to 500mA ! ?

Finally,you can test it via the arduino IDE and transpose your setup settings.

Good luck!

**Reply by Bodmer** [reply]
The display is parallel so the SPI clock speed has no effect. That line can be deleted as it will be ignored.

**Reply by BrandonStudio** [reply]
The ESP32-WROOM-32E is called a "module", whose datasheet is available on https://www.espressif.com/sites/default/files/documentation/esp32-wroom-32e_esp32-wroom-32ue_datasheet_en.pdf.

The LCD's datasheet is available on https://dl.sipeed.com/fileList/MAIX/HDK/Sipeed-LCD_Adapter_1.3&2.4_to_7/024-B2441%E8%A7%84%E6%A0%BC%E4%B9%A6%EF%BC%88%E7%9F%BD%E9%80%9F2.4%E5%AF%B8%E5%B1%8F%EF%BC%89.pdf
and its controller's (i.e. ST7789V) datasheet is available on https://newhavendisplay.com/content/datasheets/ST7789V.pdf.

IOVCC is IO's power supply.

**Reply by BrandonStudio** [reply]
If you'd like to help me further, you may refer to SiPEED specified LCD driver source code on https://github.com/sipeed/Maixduino at `/libraries/Sipeed_ST7789/` and `/cores/arduino/st7789.*`.

**Comment by TangerineDreamer** [comment]
It's a nice piece of board!
I always thougth you used differents components, maybe it comes from the connector,you can try with a meter to verify
all connections.

**Comment by BrandonStudio** [comment]
There is a good news for me that I can finally successfully run the [esp-idf example](https://github.com/espressif/esp-idf/tree/master/examples/peripherals/lcd/i80_controller) on my board with my LCD! This proves that the wire connections on my board are correct.
Now I'm going to look into the related code

**Reply by BrandonStudio** [reply]
Good luck for me T_T

**Reply by BrandonStudio** [reply]
Right now I'm able to run the example under my Arduino project (via PlatformIO). That example is fully compatible with Arduino, but you do need to separate `void app_main(void)` into `void setup(void)` and `void loop(void)` and modify some initialize parameters as the example uses the SDK in a higher version. Besides, as Arduino uses C++, you'll have to add some explicit type casts. Also, you may have to shrink the buffer size (for me, 100→50).

**Reply by BrandonStudio** [reply]
@Bodmer I saw that both SiPEED and Espressif uses DMA to send data, but it seems yours does not. May this be the problem?
